Hw04 auth

an-shem opened the pull request:
Прошу проверить 4-ю ДЗ

an-shem committed:
Сделал регистрацию пользователя

an-shem committed:
Сделал логирование и получение токена

an-shem committed:
Создал мидлвар для проверки токена

an-shem committed:
Создал ендпоинт logout для розлогирования

an-shem committed:
Создайте эндпоинт /users/current для получения данных о текущем пользователе

an-shem committed:
добавил поле owner  в схему контактов

an-shem committed:
Сделал работу с контактами приватной

Сделал так чтобы добавление и получение всех контактов было возможным только для залогиненного пользователя и работать можно было только со своими контактами

an-shem committed:
Сделал пагинацию для коллекции контактов и подкорректировал вывод данных о владельце контактов

an-shem committed:
Сделаk фильтрацию контактов по полю избранного

an-shem committed:
Сделал обновление подписки (subscription) пользователя через эндпоинт PATCH /users

an-shem committed:
Удалил проверочные console.log

KirillKulakow commented on controllers/contacts/getAll.js:
```diff
@@ -0,0 +1,28 @@
+  const { page = 1, limit = 10, favorite } = req.query;
+  const skip = (page - 1) * limit;
+
+  let contacts = null;
```
тут можна проще 
```
const filter = { owner: _id }
if (favorite) {
  filter.favorite = favorite
}
const contacts = await Contact.find(filter, '', {
       skip,
       limit: Number(limit),
     }).populate('owner', '_id email subscription');
```

an-shem replied:
исправил

KirillKulakow commented on controllers/contacts/getAll.js:
```diff
@@ -0,0 +1,28 @@
+const getAll = async (req, res) => {
+  const { _id } = req.user;
+  const { page = 1, limit = 10, favorite } = req.query;
+  const skip = (page - 1) * limit;
```
перед этой строкой было бы хорошо делать валидацию page и limit на число через isNaN

an-shem replied:
сделал валидацию

KirillKulakow commented on controllers/users/login.js:
```diff
@@ -0,0 +1,36 @@
+const login = async (req, res) => {
+  const { email, password } = req.body;
+  const user = await User.findOne({ email });
+  const comparePass = bcrypt.compareSync(password, user.password);
```
тут наверно лучше будет написать по другому
`const comparePass = bcrypt.compareSync(password, (user && user.password) || "");`

KirillKulakow replied:
если юзер не будет найден будет ошибка

an-shem replied:
подправил

KirillKulakow commented:
еще нужно внести правки в логику удаления и редактирования контакта
чтобы можно было удалить/редактировать только свой контакт
в фильтр добавить не только ид контакта но и ид собственника

KirillKulakow reviewed (changes requested):
Need fix

an-shem committed:
Внес все рекомендации по ДЗ

- поправил код в указанных местах
- сделал получение/редактирование /удаление контактов возможным только для авторизированного пользователя и только со своими контактами

an-shem commented:
Внес поправки, прошу проверить

KirillKulakow commented on controllers/contacts/getById.js:
```diff
@@ -1,12 +1,12 @@
+  const { _id } = req.user;
   const { contactId } = req.params;
-  const contact = await Contact.findById(contactId);
+  const contact = await Contact.findOne({ _id: contactId, owner: { _id: _id } });
```
эээ а почему так написал `owner: { _id: _id }`?
оно так работает?
обычно так: Contact.findOne({ _id: contactId, owner:  _id })

an-shem replied:
Работает) сам думал что вроде не правильно... Но работает

KirillKulakow commented on controllers/contacts/removeById.js:
```diff
@@ -1,12 +1,12 @@
+  const { _id } = req.user;
   const { contactId } = req.params;
-  const removedСontact = await Contact.findByIdAndRemove(contactId);
+  const removedСontact = await Contact.findOneAndRemove({ _id: contactId, owner: { _id: _id } });
```
прошлый коммент

KirillKulakow commented on controllers/contacts/updateById.js:
```diff
@@ -0,0 +1,17 @@
+const updateById = async (req, res) => {
+  const { _id } = req.user;
+  const { contactId } = req.params;
+  const updateContact = await Contact.findOneAndUpdate({ _id: contactId, owner: { _id: _id } }, req.body, {
```
прошлый коммент

KirillKulakow commented on controllers/contacts/updateFavorite.js:
```diff
@@ -1,13 +1,17 @@
   const { contactId } = req.params;
   const { favorite } = req.body;
-  const updateContact = await Contact.findByIdAndUpdate(contactId, { favorite }, { new: true });
+  const updateContact = await Contact.findOneAndUpdate(
```
прошлый коммент

KirillKulakow reviewed (changes requested):
need fix

an-shem commented:
Исправил

an-shem committed:
Внес поправки в объект поиска.

an-shem committed:
Изменил вывод информации о контакте

- Расширил информацию о владельце контакта
